src/ecdsa/ellipticcurve.py

```diff
@@ -174,12 +175,23 @@ def __init__(self, curve, x, y, z, order=None, generator=False):
             self.__y = y
             self.__z = z
             self.__order = order
+        self.__generator = generator
         self.__precompute = []
-        if generator:
+
+        # when enabled, multiplication tables are precomputed only when actually needed:
+        # that is, the specific curve is actually first used in user code. otherwise, multiplication
+        # tables for _all_ curves supported are precomputed at library _import_ time effectively
+        ENABLE_LAZY_PRECOMPUTE = os.environ.get('PYTHON_ECDSA_ENABLE_LAZY_PRECOMPUTE', None) is not None
+        if not ENABLE_LAZY_PRECOMPUTE:
+            self._maybe_precompute()
+
+    def _maybe_precompute(self):
```

Comment:
this is not thread safe, if two threads use the same object it will end up with malformed `__precompute` table; see `scale()`

Reply:
as far as I see: all call contexts of `maybe_precompute` already have critical sections

Reply:
and the original pre-computation in the ctor doesn't need one .. I think? an object cannot be accessed from 2 threads while it is still being constructed?

Reply:
technically, if you mess around `__new__` and mess with inheritance really badyly, you could end up with two threads in an `__init__` at the same time, but not in regular use, so yes, we didn't need locks before because in practice it's not possible to have two threads in `__init__` at the same time

> as far as I see: all call contexts of maybe_precompute already have critical sections

huh? as long as two threads start execution of the _maybe_precompute at the same time, and one thread checks the if before the second thread runs the first `.append()` both threads will execute the entirety of it

and even if not, then the other thread can think the __precompute list is fully initialised when it's in the process of being initialised, ending up using incomplete table